### DISCUSSION TITLE
Memory optimization

### DISCUSSION THREAD

**Discussion by jkleiber** [team401/2024-Robot-Code#130] [Ideas]
Today we were seemingly running out of memory when entering teleop. Disabling the web server seemed to help some, because we didn't see any issues for the rest of the time once we did that. I have some random additional ideas for reducing the amount of memory we use:

1. Avoid using strings if possible. They are dynamic memory hogs, instead use enums when possible. Examples
- The test mode chooser uses strings as the object, maybe we can use an enum instead in that switch (likely not a problem rn, but could be if the test mode chooser is being stored in memory).

2. Remove unused variables. We don't have many of these fortunately. Examples:
- We are not using `brake` in `CommandSwerveDrivetrain`: https://github.com/team401/2024-Robot-Code/blob/1ab50a8195bd6003cd0752270b4050cfcf78cd8a/src/main/java/frc/robot/subsystems/drive/CommandSwerveDrivetrain.java#L79

3. Avoid allocating memory to things that should only be used in sim, unless sim is active. We should make these things null if the robot is real, and not use them. Potentially big savings if these things are expensive Examples:
- `ScoringSubsystem `creates a bunch of mechanisms at the top of the class
- `Telemetry` creates a bunch of Mechanism2Ds, which most likely can be ignored in real life
- `EndgameSubsystem`

4. (This one is suspicious and might not help): Avoid using `new` in looping functions where reasonable. If the garbage collector doesn't keep up for some reason, then maybe we end up using a lot of RAM if we do this. Declaring variables up front will likely use more memory, but it would at least upper bound our memory usage. This may not be worth pursuing initially since the garbage collector should be handling this. Examples:
- Creating a new `Pose2d `every cycle of `robotPeriodic`: https://github.com/team401/2024-Robot-Code/blob/1ab50a8195bd6003cd0752270b4050cfcf78cd8a/src/main/java/frc/robot/RobotContainer.java#L666
- Creating a new `ChassisSpeeds `every time we execute `DriveWithJoysticks`: https://github.com/team401/2024-Robot-Code/blob/1ab50a8195bd6003cd0752270b4050cfcf78cd8a/src/main/java/frc/robot/commands/DriveWithJoysticks.java#L90
- Various instances in LED.java of creating `new int[3]` for colors (which isn't even relevant for our current investigation yet since LEDs are disabled.
- Speaker auto-align has a couple as arguments to the `calculateDesiredHeading`: https://github.com/team401/2024-Robot-Code/blob/1ab50a8195bd6003cd0752270b4050cfcf78cd8a/src/main/java/frc/robot/subsystems/drive/CommandSwerveDrivetrain.java#L244
- Logging the chassis speeds in `controlDrivetrain`: https://github.com/team401/2024-Robot-Code/blob/1ab50a8195bd6003cd0752270b4050cfcf78cd8a/src/main/java/frc/robot/subsystems/drive/CommandSwerveDrivetrain.java#L284
- `calculateDesiredHeading` itself has some room for optimization (we could use `double`s exclusively rather than making a bunch of new Pose2Ds (or at least re-using the Pose2Ds we have) to do this calculation. This one is a bit more complex, so we probably want to avoid changes to this for now.)
- During the logging of the 3D elevator output in `EndgameSubsystem::periodic`, which really should only be done in sim anyway: https://github.com/team401/2024-Robot-Code/blob/1ab50a8195bd6003cd0752270b4050cfcf78cd8a/src/main/java/frc/robot/subsystems/endgame/EndgameSubsystem.java#L108
- When setting the belt voltage in the IntakeIOSparkMax: https://github.com/team401/2024-Robot-Code/blob/1ab50a8195bd6003cd0752270b4050cfcf78cd8a/src/main/java/frc/robot/subsystems/intake/IntakeIOSparkMax.java#L60
- In CameraIOPhoton when calculating tag distances (this isn't relevant yet): https://github.com/team401/2024-Robot-Code/blob/1ab50a8195bd6003cd0752270b4050cfcf78cd8a/src/main/java/frc/robot/subsystems/localization/CameraIOPhoton.java#L113
- The trapezoidal profile's initial and goal states in `AimerIORoboRio` and `HoodIOSparkFlex `in `updateInputs`
- Telemetry.java

@PChild @redPlover @charlesmackenzie feel free to add your thoughts here as well

**Comment by PChild** [comment]
One thing we could maybe look at is [disabling Phoenix Diagnostic Server](https://api.ctr-electronics.com/phoenix6/release/java/com/ctre/phoenix6/unmanaged/Unmanaged.html#setPhoenixDiagnosticsStartTime(double)). This might help memory a bit but I haven't quantified how much, and it might break robot code, but it should be very fast to test.

Depending on how fast @charlesmackenzie can get it running we could also try VisualVM like mentioned in #129 to see where we should be focusing optimization efforts.

**Comment by PChild** [comment]
[We could also throw money at the problem](https://www.andymark.com/products/ni-roborio-2)

**Reply by jkleiber** [reply]
I think it's worth it to throw money at this problem given the computational requirements you need these days to have an elite control system. This would double our RAM if I understand the specs correctly 

Obviously not a solution for Blacksburg, but could we get this prior to VAFAL?

**Reply by PChild** [reply]
I think we can afford it, will set up a PO tonight and discuss.

**Reply by PChild** [reply]
RIO2 is ordered

**Comment by PChild** [comment]
[This](https://www.chiefdelphi.com/t/memory-leaks/439549/11?u=pchild) seems like it would be very easy for us to add given that we already have a USB drive installed for logging

**Reply by jkleiber** [reply]
I kinda don't want to touch the swap right before comp but yeah we could definitely try this if it seems easy and can be tested with low risk of bricking the roboRIO

**Reply by PChild** [reply]
I figure we test it on a RIO not on the bot first, if it doesn't brick and we find we need it we do it. If we don't need then don't touch.

**Reply by jkleiber** [reply]
I was literally just thinking the same thing

**Comment by PChild** [comment]
Closing this one until next year when we max out RIO2 RAM with AKit2025
